File: src/protect_client.py
"""
UniFi Protect API client — routed through the UniFi Site Manager cloud API.

All requests go to api.ui.com using the X-API-KEY header.  On first use,
the client discovers the console host ID automatically; set PROTECT_HOST_ID
in .env to pin a specific console when you have more than one.
"""

from __future__ import annotations

import base64
import logging
from dataclasses import dataclass, field

import httpx

logger = logging.getLogger(__name__)

CLOUD_BASE = "https://api.ui.com"

# Protect event types we care about
INTERESTING_TYPES = {"motion", "ring", "smartDetectZone", "smartDetectLine"}


@dataclass
class ProtectEvent:
    id: str
    type: str
    camera_id: str
    camera_name: str
    start: int  # epoch ms
    end: int | None  # epoch ms, None if still in progress
    score: int
    smart_detect_types: list[str] = field(default_factory=list)
    thumbnail_b64: str | None = None  # base64-encoded JPEG, populated after fetch

    @property
    def start_iso(self) -> str:
        import datetime
        return datetime.datetime.fromtimestamp(self.start / 1000).isoformat(timespec="seconds")

    def __repr__(self) -> str:
        label = "/".join(self.smart_detect_types) if self.smart_detect_types else self.type
        return f"<ProtectEvent {self.id} {label} cam={self.camera_name} at={self.start_iso}>"


class ProtectClient:
    """Async client for UniFi Protect, proxied through the Site Manager cloud API."""

    def __init__(self, api_key: str, host_id: str | None = None) -> None:
        self._host_id = host_id
        self._http = httpx.AsyncClient(
            base_url=CLOUD_BASE,
            headers={"X-API-KEY": api_key},
            timeout=60,  # cloud proxy adds ~800 ms latency
            follow_redirects=True,
        )
        self._cameras: dict[str, str] = {}  # id -> display name

    # ------------------------------------------------------------------
    # Host discovery
    # ------------------------------------------------------------------

    async def _resolve_host_id(self) -> None:
        """Auto-discover the console host ID from GET /v1/hosts."""
        resp = await self._http.get("/v1/hosts")
        resp.raise_for_status()
        hosts = resp.json().get("data", resp.json())  # handle both envelope shapes
        if not hosts:
            raise RuntimeError("No UniFi consoles found under this API key.")
        if len(hosts) > 1:
            ids = [h["id"] for h in hosts]
            raise RuntimeError(
                f"Multiple consoles found: {ids}\n"
                "Set PROTECT_HOST_ID in .env to choose one."
            )
        self._host_id = hosts[0]["id"]
        logger.info("Auto-discovered console host ID: %s", self._host_id)

    def _protect(self, path: str) -> str:
        """Build the cloud proxy path for a Protect API endpoint."""
        return f"/v1/connector/consoles/{self._host_id}/proxy/protect/api{path}"

    # ------------------------------------------------------------------
    # Bootstrap — load camera names once
    # ------------------------------------------------------------------

    async def load_cameras(self) -> None:
        if not self._host_id:
            await self._resolve_host_id()
        resp = await self._http.get(self._protect("/bootstrap"))
        resp.raise_for_status()
        data = resp.json()
        self._cameras = {
            cam["id"]: cam.get("name") or cam.get("id")
            for cam in data.get("cameras", [])
        }
        logger.info("Loaded %d cameras", len(self._cameras))

    # ------------------------------------------------------------------
    # Events
    # ------------------------------------------------------------------

    async def fetch_events(self, since_ms: int, until_ms: int) -> list[ProtectEvent]:
        """Return completed events whose start time falls in [since_ms, until_ms)."""
        resp = await self._http.get(
            self._protect("/events"),
            params={"start": since_ms, "end": until_ms},
        )
        resp.raise_for_status()

        events: list[ProtectEvent] = []
        for raw in resp.json():
            if raw.get("type") not in INTERESTING_TYPES:
                continue
            if raw.get("end") is None:
                continue  # skip events still in progress

            camera_id = raw.get("camera", "")
            events.append(
                ProtectEvent(
                    id=raw["id"],
                    type=raw["type"],
                    camera_id=camera_id,
                    camera_name=self._cameras.get(camera_id, camera_id),
                    start=raw["start"],
                    end=raw.get("end"),
                    score=raw.get("score", 0),
                    smart_detect_types=raw.get("smartDetectTypes", []),
                )
            )
        return events

    # ------------------------------------------------------------------
    # Thumbnails
    # ------------------------------------------------------------------

    async def fetch_thumbnail_b64(self, event_id: str) -> str | None:
        """Return a base64-encoded JPEG thumbnail, or None on failure."""
        try:
            resp = await self._http.get(
                self._protect(f"/events/{event_id}/thumbnail"),
                params={"width": 640},
            )
            if resp.status_code == 404:
                return None
            resp.raise_for_status()
            return base64.standard_b64encode(resp.content).decode()
        except httpx.HTTPError as exc:
            logger.warning("Could not fetch thumbnail for %s: %s", event_id, exc)
            return None

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    async def close(self) -> None:
        await self._http.aclose()

File: src/protectClient.ts
/**
 * UniFi Protect API client routed through the UniFi Site Manager cloud API.
 * All requests go to api.ui.com with the X-API-KEY header.
 */

import type { Logger } from 'homebridge';

const CLOUD_BASE = 'https://api.ui.com';
const INTERESTING_TYPES = new Set(['motion', 'ring', 'smartDetectZone', 'smartDetectLine']);

export interface ProtectEvent {
  id: string;
  type: string;
  cameraId: string;
  cameraName: string;
  start: number;
  end: number | null;
  score: number;
  smartDetectTypes: string[];
  thumbnailB64?: string;
}

export class ProtectClient {
  private hostId?: string;
  private readonly cameras = new Map<string, string>();
  private readonly headers: Record<string, string>;

  constructor(
    apiKey: string,
    hostId?: string,
    private readonly log?: Logger,
  ) {
    this.hostId = hostId;
    this.headers = { 'X-API-KEY': apiKey, 'Accept': 'application/json' };
  }

  private protectUrl(path: string): string {
    return `${CLOUD_BASE}/v1/connector/consoles/${this.hostId}/proxy/protect/api${path}`;
  }

  private async resolveHostId(): Promise<void> {
    const res = await fetch(`${CLOUD_BASE}/v1/hosts`, { headers: this.headers });
    if (!res.ok) throw new Error(`GET /v1/hosts → ${res.status}`);
    const json = await res.json() as { data?: unknown[] } | unknown[];
    const hosts = (Array.isArray(json) ? json : (json as { data?: unknown[] }).data ?? []) as Array<{ id: string }>;
    if (!hosts.length) throw new Error('No UniFi consoles found under this API key.');
    if (hosts.length > 1) {
      const ids = hosts.map(h => h.id).join('\n  ');
      throw new Error(`Multiple consoles found — set hostId in Homebridge config:\n  ${ids}`);
    }
    this.hostId = hosts[0].id;
    this.log?.info('Auto-discovered console host ID: %s', this.hostId);
  }

  async loadCameras(): Promise<void> {
    if (!this.hostId) await this.resolveHostId();
    const res = await fetch(this.protectUrl('/bootstrap'), { headers: this.headers });
    if (!res.ok) throw new Error(`Bootstrap → ${res.status}`);
    const data = await res.json() as { cameras?: Array<{ id: string; name?: string }> };
    for (const cam of data.cameras ?? []) {
      this.cameras.set(cam.id, cam.name ?? cam.id);
    }
    this.log?.info('Loaded %d camera(s)', this.cameras.size);
  }

  getCameras(): Map<string, string> {
    return this.cameras;
  }

  async fetchEvents(sinceMs: number, untilMs: number): Promise<ProtectEvent[]> {
    const url = new URL(this.protectUrl('/events'));
    url.searchParams.set('start', String(sinceMs));
    url.searchParams.set('end', String(untilMs));
    const res = await fetch(url.toString(), { headers: this.headers });
    if (!res.ok) throw new Error(`Events → ${res.status}`);
    const raw = await res.json() as Array<Record<string, unknown>>;
    return raw
      .filter(e => INTERESTING_TYPES.has(e['type'] as string) && e['end'] != null)
      .map(e => ({
        id: e['id'] as string,
        type: e['type'] as string,
        cameraId: (e['camera'] as string) ?? '',
        cameraName: this.cameras.get((e['camera'] as string) ?? '') ?? (e['camera'] as string) ?? '',
        start: e['start'] as number,
        end: (e['end'] as number) ?? null,
        score: (e['score'] as number) ?? 0,
        smartDetectTypes: (e['smartDetectTypes'] as string[]) ?? [],
      }));
  }

  async fetchThumbnailB64(eventId: string): Promise<string | undefined> {
    const url = new URL(this.protectUrl(`/events/${eventId}/thumbnail`));
    url.searchParams.set('width', '640');
    try {
      const res = await fetch(url.toString(), { headers: this.headers });
      if (!res.ok) return undefined;
      const buf = await res.arrayBuffer();
      return Buffer.from(buf).toString('base64');
    } catch {
      return undefined;
    }
  }
}
